From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Esme Povirk <esme@codeweavers.com>
Date: Wed, 6 Oct 2021 15:44:57 -0500
Subject: [PATCH] setupapi: Compare dll files before overwriting them.

This is mainly to prevent us from clobbering reflinks created by
the proton script, and to make sure dll symlinks (which we can't
overwrite) are still registered properly.

CW-Bug-Id: #18633
Valve-commit: 13d0701f1715567b51a4676e9a22b33cef018cf0
Picked: clean
Class: valve
Evidence: docs/evidence/2026-09-28-wine-proton-rebase.md
Offered-upstream: no
---
 dlls/setupapi/fakedll.c | 153 ++++++++++++++++++++++++++++++++++++----
 1 file changed, 139 insertions(+), 14 deletions(-)

diff --git a/dlls/setupapi/fakedll.c b/dlls/setupapi/fakedll.c
index 4dce22b..7247306 100644
--- a/dlls/setupapi/fakedll.c
+++ b/dlls/setupapi/fakedll.c
@@ -428,7 +428,7 @@ static const WCHAR *enum_load_path( unsigned int idx )
 }
 
 /* try to load a pre-compiled fake dll */
-static void *load_fake_dll( const WCHAR *name, SIZE_T *size )
+static void *load_fake_dll( const WCHAR *name, SIZE_T *size, WCHAR **out_filename )
 {
     const WCHAR *build_dir = _wgetenv( L"WINEBUILDDIR" );
     const WCHAR *path;
@@ -476,11 +476,94 @@ static void *load_fake_dll( const WCHAR *name, SIZE_T *size )
     }
 
 done:
+    if (res == 1)
+    {
+        *out_filename = malloc( (wcslen( ptr ) + 1) * sizeof(WCHAR) );
+        wcscpy( *out_filename, ptr );
+        free( file );
+        return data;
+    }
     free( file );
-    if (res == 1) return data;
     return NULL;
 }
 
+/* Check if the 2 files are already the same */
+static BOOL fake_dll_matches( const WCHAR *source, const WCHAR *dest )
+{
+    WIN32_FILE_ATTRIBUTE_DATA data1, data2;
+    HANDLE h;
+    char *buf;
+    DWORD size, bytesread;
+    BOOL result;
+
+    if (!GetFileAttributesExW( source, GetFileExInfoStandard, &data1 ))
+        return FALSE;
+
+    if (!GetFileAttributesExW( dest, GetFileExInfoStandard, &data2 ))
+        return FALSE;
+
+    if ((data2.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) == FILE_ATTRIBUTE_REPARSE_POINT)
+        /* If it's a symlink, assume it was put there by the proton script and is therefore correct */
+        return TRUE;
+
+    if (data1.nFileSizeHigh != 0 ||
+        data2.nFileSizeHigh != 0 ||
+        data1.nFileSizeLow != data2.nFileSizeLow)
+    {
+        return FALSE;
+    }
+
+    size = data1.nFileSizeLow;
+
+    /* If size and mtime matches, don't bother comparing contents */
+    if ((data1.ftLastWriteTime.dwLowDateTime != 0 || data1.ftLastWriteTime.dwHighDateTime != 0) &&
+        data1.ftLastWriteTime.dwLowDateTime == data2.ftLastWriteTime.dwLowDateTime &&
+        data1.ftLastWriteTime.dwHighDateTime == data2.ftLastWriteTime.dwHighDateTime)
+        return TRUE;
+
+    buf = malloc( size * 2 );
+
+    if (!buf)
+        return FALSE;
+
+    h = CreateFileW( source, GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL );
+    if (h == INVALID_HANDLE_VALUE)
+    {
+        free( buf );
+        return FALSE;
+    }
+
+    if (!ReadFile( h, buf, size, &bytesread, NULL ) || bytesread != size)
+    {
+        CloseHandle( h );
+        free( buf );
+        return FALSE;
+    }
+
+    CloseHandle( h );
+
+    h = CreateFileW( dest, GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL );
+    if (h == INVALID_HANDLE_VALUE)
+    {
+        free( buf );
+        return FALSE;
+    }
+
+    if (!ReadFile( h, buf + size, size, &bytesread, NULL ) || bytesread != size)
+    {
+        CloseHandle( h );
+        free( buf );
+        return FALSE;
+    }
+
+    CloseHandle( h );
+
+    result = !memcmp( buf, buf + size, size );
+
+    free( buf );
+    return result;
+}
+
 /* create the fake dll destination file */
 static HANDLE create_dest_file( const WCHAR *name, BOOL delete )
 {
@@ -958,7 +1041,11 @@ static int install_fake_dll( WCHAR *dest, WCHAR *file, BOOL delete, struct list
     destname[len] = 0;
     if (!add_handled_dll( destname )) ret = -1;
 
-    if (ret != -1)
+    if (ret != -1 && fake_dll_matches( file, dest ))
+    {
+        register_fake_dll( dest, data, size, delay_copy );
+    }
+    else if (ret != -1)
     {
         HANDLE h = create_dest_file( dest, delete );
 
@@ -990,6 +1077,13 @@ static void delay_copy_files( struct list *delay_copy )
     LIST_FOR_EACH_ENTRY_SAFE( copy, next, delay_copy, struct delay_copy, entry )
     {
         list_remove( &copy->entry );
+
+        if ( fake_dll_matches( copy->src, copy->dest ) )
+        {
+            free( copy );
+            continue;
+        }
+
         ret = read_file( copy->src, &data, &size );
         if (ret != 1)
         {
@@ -1101,6 +1195,7 @@ BOOL create_fake_dll( const WCHAR *name, const WCHAR *source )
     BOOL ret = FALSE;
     SIZE_T size;
     const WCHAR *filename;
+    WCHAR *source_filename;
     void *buffer;
     BOOL delete = !wcscmp( source, L"-" );  /* '-' source means delete the file */
 
@@ -1117,24 +1212,54 @@ BOOL create_fake_dll( const WCHAR *name, const WCHAR *source )
 
     add_handled_dll( filename );
 
-    if (!(h = create_dest_file( name, delete ))) return TRUE;  /* not a fake dll */
-    if (h == INVALID_HANDLE_VALUE) return FALSE;
+    if (delete)
+    {
+        if (!(h = create_dest_file( name, delete ))) return TRUE;  /* not a fake dll */
+        if (h == INVALID_HANDLE_VALUE) return FALSE;
 
-    if (!wcscmp( source, L"*" ))  /* '*' source means create fake file for a non-existent dll */
+        /* '-' source means delete the file */
+        TRACE( "deleting %s\n", debugstr_w(name) );
+        ret = FALSE;
+
+        CloseHandle( h );
+        DeleteFileW( name );
+    }
+    else if (!wcscmp( source, L"*" ))  /* '*' source means create fake file for a non-existent dll */
     {
+        if (!(h = create_dest_file( name, delete ))) return TRUE;  /* not a fake dll */
+        if (h == INVALID_HANDLE_VALUE) return FALSE;
+
         ret = build_fake_dll( h, name );
+
+        CloseHandle( h );
+        if (!ret) DeleteFileW( name );
     }
-    else if ((buffer = load_fake_dll( source, &size )))
+    else if ((buffer = load_fake_dll( source, &size, &source_filename )))
     {
-        DWORD written;
+        if (fake_dll_matches( source_filename, name ))
+        {
+            free( source_filename );
 
-        ret = (WriteFile( h, buffer, size, &written, NULL ) && written == size);
-        if (ret) register_fake_dll( name, buffer, size, &delay_copy );
-        else ERR( "failed to write to %s (error=%lu)\n", debugstr_w(name), GetLastError() );
-    }
+            register_fake_dll( name, buffer, size, &delay_copy );
+            ret = TRUE;
+        }
+        else
+        {
+            DWORD written;
 
-    CloseHandle( h );
-    if (!ret) DeleteFileW( name );
+            free( source_filename );
+
+            if (!(h = create_dest_file( name, FALSE ))) return TRUE;  /* not a fake dll */
+            if (h == INVALID_HANDLE_VALUE) return FALSE;
+
+            ret = (WriteFile( h, buffer, size, &written, NULL ) && written == size);
+            if (ret) register_fake_dll( name, buffer, size, &delay_copy );
+            else ERR( "failed to write to %s (error=%lu)\n", debugstr_w(name), GetLastError() );
+
+            CloseHandle( h );
+            if (!ret) DeleteFileW( name );
+        }
+    }
 
     delay_copy_files( &delay_copy );
     return ret;
